internal/pack/pack.go

```diff
@@ -170,26 +170,74 @@ func (p *Packer) String() string {
 	return fmt.Sprintf("<Packer %d blobs, %d bytes>", len(p.blobs), p.bytes)
 }
 
-const maxHeaderSize = 16 * 1024 * 1024
+var (
+	// size of the header-length field at the end of the file
+	headerLengthSize = binary.Size(uint32(0))
+	// we require at least one entry in the header, and one blob for a pack file
+	minFileSize = entrySize + crypto.Extension + uint(headerLengthSize)
+)
+
+const (
+	maxHeaderSize = 16 * 1024 * 1024
+	// number of header enries to download as part of header-length request
+	eagerEntries = 15
+)
+
+// readRecords reads count records from the underlying ReaderAt, returning the
+// raw header, the total number of records in the header, and any error.  If
+// the header contains fewer than count entries, the return value is truncated.
+func readRecords(rd io.ReaderAt, size int64, count int) ([]byte, int, error) {
+	var bufsize int
+	bufsize += count * int(entrySize)
+	bufsize += crypto.Extension
+	bufsize += headerLengthSize
+
+	if bufsize > int(size) {
+		bufsize = int(size)
+	}
 
-// size of the header-length field at the end of the file
-var headerLengthSize = binary.Size(uint32(0))
+	b := make([]byte, bufsize)
+	off := size - int64(bufsize)
+	if _, err := rd.ReadAt(b, off); err != nil {
+		return nil, 0, err
+	}
+
+	header := b[len(b)-headerLengthSize:]
+	b = b[:len(b)-headerLengthSize]
+	hl := binary.LittleEndian.Uint32(header)
+	debug.Log("header length: %v", hl)
+
+	var err error
+	switch {
+	case hl == 0:
+		err = InvalidFileError{Message: "header length is zero"}
+	case hl < crypto.Extension:
+		err = InvalidFileError{Message: "header length is too small"}
+	case (hl-crypto.Extension)%uint32(entrySize) != 0:
+		err = InvalidFileError{Message: "header length is invalid"}
+	case int64(hl) > size-int64(headerLengthSize):
+		err = InvalidFileError{Message: "header is larger than file"}
+	case int64(hl) > maxHeaderSize:
+		err = InvalidFileError{Message: "header is larger than maxHeaderSize"}
+	}
+	if err != nil {
+		return nil, 0, errors.Wrap(err, "readHeader")
+	}
 
-// we require at least one entry in the header, and one blob for a pack file
-var minFileSize = entrySize + crypto.Extension + uint(headerLengthSize)
+	c := (int(hl) - crypto.Extension) / int(entrySize)
```

Comment:
maybe rename `c` to `entryCount`?

Reply:
I think `c` is appropriate, since it only lives for like 5 lines.

Reply:
Maybe `total` to capture the meaning of that field.